File: server/Models/MarkdownEmail.java
package server.Models;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.commonmark.node.Node;
import org.commonmark.parser.Parser;
import org.commonmark.renderer.html.HtmlRenderer;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;
import org.yaml.snakeyaml.nodes.Tag;
import org.yaml.snakeyaml.representer.Representer;

public class MarkdownEmail {
    private String markdown;
    private String html;
    private Map<String, Object> data;
    private List<String> tags;

    public MarkdownEmail() {
        this.tags = new ArrayList<>();
    }

    public static MarkdownEmail fromFile(String path) throws IOException {
        MarkdownEmail email = new MarkdownEmail();
        email.markdown = readFile(path);
        email.render();
        return email;
    }

    public static MarkdownEmail fromString(String markdown) {
        MarkdownEmail email = new MarkdownEmail();
        email.markdown = markdown;
        email.render();
        return email;
    }

    public boolean isValid() {
        if (data == null) {
            return false;
        }
        return data.containsKey("Subject") && data.containsKey("Summary");
    }

    private void render() {
        if (markdown == null) {
            throw new IllegalStateException("Markdown is null; be sure to set that first");
        }

        Parser parser = Parser.builder().build();
        Node document = parser.parse(markdown);
        HtmlRenderer renderer = HtmlRenderer.builder().build();
        html = renderer.render(document);

        Yaml yaml = new Yaml(new Constructor(), new Representer(), new DumperOptions(), new Tag("!"));
        data = yaml.load(new StringReader(markdown));

        if (!data.containsKey("Slug")) {
            data.put("Slug", ((String) data.get("Subject")).toLowerCase().replace(" ", "-"));
        }
        if (!data.containsKey("SendToTag")) {
            data.put("SendToTag", "*");
        }
    }

    private static String readFile(String path) throws IOException {
        StringBuilder content = new StringBuilder();
        try (FileReader reader = new FileReader(new File(path))) {
            char[] buffer = new char[1024];
            int bytesRead;
            while ((bytesRead = reader.read(buffer)) != -1) {
                content.append(buffer, 0, bytesRead);
            }
        }
        return content.toString();
    }

    // Getters and setters
    public String getMarkdown() {
        return markdown;
    }

    public void setMarkdown(String markdown) {
        this.markdown = markdown;
    }

    public String getHtml() {
        return html;
    }

    public void setHtml(String html) {
        this.html = html;
    }

    public Map<String, Object> getData() {
        return data;
    }

    public void setData(Map<String, Object> data) {
        this.data = data;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }
}

File: server/Models/MarkdownEmail.cs
using System.Dynamic;
using Markdig;
using YamlDotNet.Core;
using YamlDotNet.Core.Events;
using YamlDotNet.Serialization;
using Tailwind.Mail.Services;

namespace Tailwind.Mail.Models;

public class MarkdownEmail{
  public string? Markdown { get; set; }
  public string? Html { get; set; }
  /// <summary>
  /// Gets the fully accessible HTML email with proper structure, ARIA attributes,
  /// and live regions for screen reader support.
  /// </summary>
  public string? AccessibleHtml { get; set; }
  public dynamic? Data { get; set; }
  public List<string> Tags { get; set; } = new List<string>();
  /// <summary>
  /// When true, uses the accessible HTML wrapper template.
  /// Set to false to use raw HTML without accessibility enhancements.
  /// </summary>
  public bool UseAccessibleTemplate { get; set; } = true;
  public MarkdownEmail()
  {

  }
  public static MarkdownEmail FromFile(string path){
    var email = new MarkdownEmail();
    email.Markdown = File.ReadAllText(path);
    email.Render();
    return email;
  } 
  public static MarkdownEmail FromString(string markdown){
    var email = new MarkdownEmail();
    email.Markdown = markdown;
    email.Render();
    return email;
  }
  public bool IsValid(){
    if(Data == null) return false;
    return Data.Subject != null && Data.Summary != null;
  }
  private void Render(){
    if(Markdown == null){
      throw new Exception("Markdown is null; be sure to set that first");
    }
    var pipeline = new MarkdownPipelineBuilder().UseAdvancedExtensions().Build();
    Html = Markdig.Markdown.ToHtml(Markdown, pipeline);

    //data
    var yamler = new DeserializerBuilder().IgnoreUnmatchedProperties().Build();

    using (var input = new StringReader(Markdown))
    {
      var parser = new Parser(input);
      parser.Consume<StreamStart>();
      parser.Consume<DocumentStart>();
      //dyamic acti0n
      Data = yamler.Deserialize<ExpandoObject>(parser);
      parser.Consume<DocumentEnd>();
    }
    var expando = (IDictionary<string, object>)Data;
    if(!expando.ContainsKey("Slug")){
      Data.Slug = Data.Subject.ToLower().Replace(" ", "-");
    }
    if(!expando.ContainsKey("SendToTag")){
      Data.SendToTag = "*"; //send to everyone
    }
    
    // Generate accessible HTML with screen reader support and live regions
    if(UseAccessibleTemplate && Html != null && Data != null){
      var subject = Data.Subject?.ToString() ?? "";
      var summary = Data.Summary?.ToString() ?? "";
      // Get language from frontmatter or default to "en"
      var lang = expando.ContainsKey("Lang") ? Data.Lang?.ToString() ?? "en" : "en";
      AccessibleHtml = AccessibleEmailTemplate.Wrap(subject, summary, Html, lang);
    } else {
      AccessibleHtml = Html;
    }
  }
}